File: code_advent/year_2015/day_03.py
""" http://adventofcode.com/2015/day/3 """
from collections import namedtuple

Point = namedtuple('Point', ['x', 'y'])


def move(point, direction):
    """Return new position based on direction"""
    if direction == '^':
        return Point(point.x, point.y + 1)
    if direction == 'v':
        return Point(point.x, point.y - 1)
    if direction == '<':
        return Point(point.x - 1, point.y)
    return Point(point.x + 1, point.y)


def get_houses(directions):
    """Return total count of houses touched in series of directions"""
    position = Point(0, 0)
    houses = [position]
    for direction in directions:
        position = move(position, direction)
        houses.append(position)
    return len(set(houses))

File: elixir/elixir_advent/lib/year_2015/day_03.ex
# 2015 - Day 03: http://adventofcode.com/2015/day/3

defmodule ElixirAdvent.Year2015.Day03 do
  def part1(input) do
    directions = directions_to_list(ElixirAdvent.read_input(input))

    reduced = Enum.reduce(directions, [[0, 0], MapSet.new([[0, 0]])], fn direction, acc ->
      current_position = move(Enum.at(acc, 0), direction)
      [current_position, MapSet.put(Enum.at(acc, 1), current_position)]
    end)

    houses = Enum.at(reduced, 1)
    MapSet.size(houses)
  end

  def part2(input) do
    ElixirAdvent.read_input(input)
  end

  def move(position, direction) do
    case (direction) do
      ">" -> [Enum.at(position, 0) + 1, Enum.at(position, 1)]
      "<" -> [Enum.at(position, 0) - 1, Enum.at(position, 1)]
      "^" -> [Enum.at(position, 0), Enum.at(position, 1) + 1]
      "v" -> [Enum.at(position, 0), Enum.at(position, 1) - 1]
    end
  end

  def directions_to_list(input) do
    List.delete(String.split(input, "", trim: True), "")
  end
end
